File: java-algorithm/src/main/java/com/leetcode/Problem1706.java
package com.leetcode;

/**
 * 1706. Where Will the Ball Fall
 * https://leetcode.com/problems/where-will-the-ball-fall/
 */
public class Problem1706 {
    public int[] findBall(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            int r = 0, c = i;
            while (r < m) {
                if (grid[r][c] == 1) {
                    if (c == n - 1 || grid[r][c + 1] == -1) {
                        ans[i] = -1;
                        break;
                    }
                    c++;
                } else {
                    if (c == 0 || grid[r][c - 1] == 1) {
                        ans[i] = -1;
                        break;
                    }
                    c--;
                }
                r++;
            }
            if (r == m) {
                ans[i] = c;
            }
        }
        return ans;
    }
}

File: python-algo/leetcode/problem_1706.py
"""1706. Where Will the Ball Fall
https://leetcode.com/problems/where-will-the-ball-fall/
"""
from typing import List


def find_ball(grid: List[List[int]]) -> List[int]:
    m, n = len(grid), len(grid[0])
    ans = [0] * n
    for i in range(n):
        row, col = 0, i
        while row < m:
            if grid[row][col] == 1:
                if col == n - 1 or grid[row][col + 1] == -1:
                    ans[i] = -1
                    break
                col += 1
            else:
                if col == 0 or grid[row][col - 1] == 1:
                    ans[i] = -1
                    break
                col -= 1
            row += 1
        if row >= m:
            ans[i] = col
    return ans
